File: src/auto48/models/listing.py
"""Listing aggregate: a vehicle offered for sale by a seller."""

from __future__ import annotations

import enum
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import (
    BigInteger,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from auto48.db import Base

if TYPE_CHECKING:
    from auto48.models.photo import Photo
    from auto48.models.vehicle import Vehicle


class ListingStatus(enum.Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    SOLD = "sold"
    EXPIRED = "expired"


class Listing(Base):
    __tablename__ = "listings"

    id: Mapped[int] = mapped_column(primary_key=True)
    seller_id: Mapped[int] = mapped_column(ForeignKey("seller_profiles.id"), index=True)
    vehicle_id: Mapped[int] = mapped_column(ForeignKey("vehicles.id"), index=True)
    title: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text, default=None)
    # Money stored as integer EUR cents (project invariant).
    price_eur_cents: Mapped[int] = mapped_column(BigInteger, index=True)
    mileage_km: Mapped[int | None] = mapped_column(Integer, default=None)
    location_county: Mapped[str | None] = mapped_column(String(64), default=None)
    # Plain floats now; PostGIS Point later.
    lat: Mapped[float | None] = mapped_column(Float, default=None)
    lon: Mapped[float | None] = mapped_column(Float, default=None)
    status: Mapped[ListingStatus] = mapped_column(
        Enum(
            ListingStatus,
            name="listing_status",
            values_callable=lambda e: [m.value for m in e],
        ),
        default=ListingStatus.DRAFT,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    vehicle: Mapped[Vehicle] = relationship(back_populates="listings")
    photos: Mapped[list[Photo]] = relationship(
        back_populates="listing", order_by="Photo.position"
    )

File: frontend/app/types/listing.ts
/**
 * TypeScript interfaces matching the auto48 v1 API contract.
 * All money values are integer cents (EUR). Use formatEur() for display.
 */

// Values MUST match the backend enums in src/auto48/models/vehicle.py.
export type FuelType = 'petrol' | 'diesel' | 'hybrid' | 'plugin_hybrid' | 'electric' | 'lpg' | 'cng' | 'other'
export type BodyType = 'sedan' | 'hatchback' | 'wagon' | 'suv' | 'coupe' | 'convertible' | 'minivan' | 'pickup' | 'van' | 'other'
export type TransmissionType = 'manual' | 'automatic' | 'semi_automatic' | 'cvt'
export type DrivetrainType = 'fwd' | 'rwd' | 'awd'
export type ListingStatus = 'draft' | 'active' | 'sold' | 'expired'

/** Nested vehicle object as returned within a ListingResponse. */
export interface Vehicle {
  id: number
  vin?: string
  plate?: string
  make: string
  model: string
  variant?: string
  year: number
  fuel: FuelType
  body: BodyType
  transmission: TransmissionType
  drivetrain?: DrivetrainType
  specs?: Record<string, unknown>
}

/** A single photo attached to a listing. `url` is a CDN-absolute URL. */
export interface Photo {
  id: number
  listing_id: number
  url: string
  position: number
  processed: boolean
  created_at: string
}

/**
 * Full listing as returned by GET /v1/listings/{id} and items in GET /v1/listings.
 * Vehicle specs are in the nested `vehicle` object; photos are NOT embedded —
 * fetch separately via GET /v1/listings/{id}/photos.
 */
export interface Listing {
  id: number
  seller_id: number
  vehicle_id: number
  title: string
  description?: string
  /** Price in integer cents EUR */
  price_eur_cents: number
  mileage_km?: number | null
  location_county?: string
  lat?: number
  lon?: number
  status: ListingStatus
  vehicle: Vehicle
  /** ISO 8601 datetime */
  created_at: string
  /** ISO 8601 datetime */
  updated_at: string
}

/** Paginated envelope used by all list endpoints */
export interface Page<T> {
  items: T[]
  total: number
  limit: number
  offset: number
}

/** Query parameters for GET /v1/listings */
export interface ListingsQuery {
  make?: string
  model?: string
  year_min?: number
  year_max?: number
  /** Price filters are in cents */
  price_min?: number
  price_max?: number
  mileage_max?: number
  fuel?: FuelType
  body?: BodyType
  transmission?: TransmissionType
  /** Plain location string, NOT location_county */
  location?: string
  q?: string
  sort?: 'newest' | 'price_asc' | 'price_desc' | 'year_desc' | 'mileage_asc'
  limit?: number
  offset?: number
}

/** Result from GET /v1/vehicles/lookup?plate= or ?vin= */
export interface VehicleLookup {
  make?: string
  model?: string
  variant?: string
  year?: number
  fuel?: FuelType
  body?: BodyType
  transmission?: TransmissionType
  drivetrain?: DrivetrainType
  engine_cc?: number
  power_kw?: number
  color?: string
  first_registered?: string
}

/** Nested vehicle sub-object for POST /v1/listings */
export interface VehicleCreate {
  make: string
  model: string
  variant?: string
  year: number
  fuel: FuelType
  body: BodyType
  transmission: TransmissionType
  drivetrain?: DrivetrainType
  vin?: string
  plate?: string
  specs?: Record<string, unknown>
}

/** Payload for POST /v1/listings */
export interface ListingCreate {
  seller_id: number
  vehicle: VehicleCreate
  title: string
  description?: string
  /** Price in integer cents EUR */
  price_eur_cents: number
  mileage_km?: number | null
  location_county?: string
  lat?: number
  lon?: number
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Format integer cents as a localised EUR price string.
 * e.g. formatEur(1549900) → "15 499 €"
 */
export function formatEur(cents: number): string {
  return (cents / 100).toLocaleString('et-EE', {
    style: 'currency',
    currency: 'EUR',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  })
}

/**
 * Format mileage with locale-appropriate thousands separator.
 * Returns "—" when null/undefined.
 */
export function formatMileage(km: number | null | undefined): string {
  if (km == null) return '—'
  return km.toLocaleString('et-EE') + ' km'
}
